validphys2/src/validphys/n3fit_data.py

```diff
@@ -274,18 +332,38 @@ def fitting_data_dict(
         # prepare a masking rotation
         dt_trans_tr = dt_trans[tr_mask]
         dt_trans_vl = dt_trans[vl_mask]
+
+        # TODO: check the effect of this when diagonalization
+        invcovmat_tr[data_zero_tr] = 0.0
+        invcovmat_vl[data_zero_vl] = 0.0
     else:
         covmat_tr = covmat[tr_mask].T[tr_mask]
-        invcovmat_tr = np.linalg.inv(covmat_tr)
-
         covmat_vl = covmat[vl_mask].T[vl_mask]
+
+        # Remove possible correlations for 1-point datasets
+        # that should've been masked out
+        covmat_tr[data_zero_tr, :] = covmat_tr[:, data_zero_tr] = 0.0
+        covmat_vl[data_zero_vl, :] = covmat_vl[:, data_zero_vl] = 0.0
+        # Avoid infinities
+        covmat_tr[data_zero_tr, data_zero_tr] = 1.0
+        covmat_vl[data_zero_vl, data_zero_vl] = 1.0
+
+        invcovmat_tr = np.linalg.inv(covmat_tr)
         invcovmat_vl = np.linalg.inv(covmat_vl)
 
-    ndata_tr = np.count_nonzero(tr_mask)
-    expdata_tr = expdata[tr_mask].reshape(1, ndata_tr)
+        # Set to 0 the points in the diagonal that were left as 1
+        invcovmat_tr[np.ix_(data_zero_tr, data_zero_tr)] = 0.0
+        invcovmat_vl[np.ix_(data_zero_vl, data_zero_vl)] = 0.0
```

Comment:
I don't think this is mathematically sound. You take the inverse of the entire matrix, not element-by-element, so doing the inverse and then setting to 0 the points you don't want is different from slicing and then taking the inverse. The normalisation (determinant) will account for those points you effectively remove in these last two lines

Reply:
Ah, right, yes that should be fine indeed